### PR TITLE
Improve size optimizations when compiling material

### PR DESCRIPTION
**Pull request by romainguy**
This changes the behavior of the size optimizer in matc (-S), but only for GLSL and MSL. With this change we gain a ~65% size reduction on a lit material compiled for OpenGL. To get those gains we generate extra SPIRV debug information to preserve variable names and better utilize the line dictionary. Unfortunately this break the SPIRV optimizer so we skip it and instead rely on a simple DCE pass provided by glslang. We also enhance the whitespace removal pass of the GLSL minifier to move lone { and } to the previous line, which avoids generating an extra index in each shader variant. Each index being at least as big as the character itself, this is quite wasteful.

When generating SPIRV for Vulkan, we rely on spirv-opt for size optimizations as before.

### CODE REVIEW

File: libs/filamat/src/ShaderMinifier.cpp [hunk 1]
```diff
@@ -137,7 +137,7 @@ namespace {
  * - Remove leading white spaces at the beginning of each line
  * - Remove empty lines
  */
-std::string ShaderMinifier::removeWhitespace(const std::string& s) const {
+std::string ShaderMinifier::removeWhitespace(const std::string& s, bool mergeBraces) const {
```
Comment (pixelflinger) [inline libs/filamat/src/ShaderMinifier.cpp:140]: why not doing this regardless of the optimization mode?

File: NEW_RELEASE_NOTES.md [hunk 2]
```diff
@@ -8,3 +8,5 @@ appropriate header in [RELEASE_NOTES.md](./RELEASE_NOTES.md).
 
 ## Release notes for next branch cut
 
+- materials: improved size reduction of OpenGL/Metal shaders by ~65% when compiling materials with
+             size optimizations (`matc -S`)
```
Comment (pixelflinger) [inline NEW_RELEASE_NOTES.md:12]: kind of obvious, but you could add a `[recompile materials]`

**Review by pixelflinger — APPROVED** [review]

File: libs/filamat/src/ShaderMinifier.cpp (hunk 1 above)
Comment (romainguy) [reply libs/filamat/src/ShaderMinifier.cpp:140]: Because it actually makes things worse for the other cases.

File: NEW_RELEASE_NOTES.md (hunk 2 above)
Comment (romainguy) [reply NEW_RELEASE_NOTES.md:12]: Sure

**Pull request merged**